### PR TITLE
How to set up and Build a Vue app with a Flask backend using Docker

### LINKED ISSUE #3337: How to set up and Build an Vue app with a Flask and MySQL backend using Docker
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article higher ranking.*

How to set up and Build a Vue app with a Flask and MySQL backend using Docker

### Introduction paragraph (2-3 paragraphs):
Please write the Introductory paragraph(s), that would be included in your article. We will use this writing snippet to help us assess overall quality before approval. We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.
Take your time and write the content as you would intend to get it published. 

In my recent [article](https://github.com/section-engineering-education/engineering-education/pull/2974), I used React.js with Node.js, MySQL, and Docker to spin up a full-stack web application. React.js serves as a frontend javascript framework. Vue falls in the same category as React.js and Angular.js. This article will use Vue together with Flask (a Python backend framework) and MySQL to build a web application and spin the whole application stack with Docker. 

### Key takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article? Use this area to get your ideas down on the bulk of your article or tutorial.

- How to set up a basic Vue application
- Build a simple backend application using Vue 
- Set up the backend services using Flask and connect to MySQL
- Spin the application using Docker

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by mosesreigns**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [ ] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #3337

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
plagiarism check done ✅

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 1]
```diff
@@ -0,0 +1,953 @@
...
+
+### Prerequisites
+To follow along in this article, it is important to have the following:
+
+- [Python](https://www.python.org/) installed on your computer.
+- Basic working knowledge with Flask.
+- [Node.js](https://nodejs.org/en/) installed on your computer.
+- Basic working knowledge with Vue js.
+- [Docker](https://docs.docker.com/get-docker/) installed on your computer.
+
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:21]: Also include articles published on EngEd for further guidance.

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+### Overview
+- [Prerequisites](#prerequisites)
+- [Overview](#overview)
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:24]: Remove these list items

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+```
+
+#### Installing the packages
+We will use the following packages:
+
+- *Flask*: The framework providing the architectural setup for the application.
+- *Flask-sqlalchemy*: Provides useful defaults and extra helpers that make it easier to accomplish database tasks.
+- *Flask-marshmallow*: Thin integration layer for Flask useful when building APIs.
+- *Marshmallow-sqlalchemy*: Binder for sqlalchemy and marshmallow.
+- *Flask-cors*: For handling cross-origin resource access.
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:71]: Provide links to the official documentation for each of these packages.

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+```python
+class TodoSchema(ma.Schema):
+    class Meta:
+        fields = ('id','title','description')
+```
+
+From above, we define that for every todo, we will be interested in the id, title, and description.
+
+- Initialize the schema.
+
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:158]: Rewrite these and similar items as a nested headers, using h4 (####).

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 5]
```diff
@@ -0,0 +1,953 @@
...
+    }
+</script>
+```
+
+From above, we are setting the dynamic classes for the navigation bar.
+
+- Edit the style as follows:
+
+```css
+<style>
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:400]: Use the correct indentation for CSS code in this and other code blocks.

**Review by geoffrey45 — COMMENTED** [review]
Hello @mosesreigns  Please check the comments below.

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 6]
```diff
@@ -0,0 +1,950 @@
...
+
+- Start by importing packages:
+
+```python
+from flask import Flask,request, jsonify
+from flask_cors import CORS,cross_origin
+```
+
+*request* will be used to get the payload (data sent), whereas *jsonify* will be used to return JSON data. *CORS* and *cross_origin* for setting up the access policy.
+
```
[the remark this answers is not in the document]
Comment (mosesreigns) [reply content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md]: > inline code blocks

@geoffrey45 I have replaced all relevant `*` with ````

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 7]
```diff
@@ -0,0 +1,952 @@
...
+    title = request.json['title']
+    description = request.json['description']
+
+    # Create an instance
+    new_todo = Todo(title, description)
+
+    # Save the todo in the db
+    db.session.add(new_todo)
+    db.session.commit()
+
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:205]: You have not created the tables in the database yet. ie. `db.create_all()`. First create the tables for the app to work.

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 8]
```diff
@@ -0,0 +1,952 @@
...
+    }
+}
+};
+</script>
+```
+
+From the above script, we are exporting data from the component and a method that handles validation and data submission on the submission of the form.
+
+- Create an `AddTodo.vue` file and add in the following.
+
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:649]: Did you mean `EditTodo.vue`? You have already created this file.

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 9]
```diff
@@ -0,0 +1,952 @@
...
+```javascript
+// create a vuerouter instance
+const router = new VueRouter({
+mode: 'history',
+base: __dirname,
+routes: [
+    { path: '/', component: Todos,name:'home' },
+    { path: '/add-todo', component: AddTodo,name:'add-todo' },
+    { path: '/edit-todo/:id', component: EditTodo,name:'edit-todo' },
+]
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:780]: You have not imported these components.

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 10]
```diff
@@ -0,0 +1,952 @@
...
+
+# Working directory
+WORKDIR /app
+
+# Copy the dependencies
+COPY requirements.txt requirements.txt
+
+# Install the dependencies
+RUN pip3 install -r requirements.txt
+
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:836]: You are working with `requirements.txt` but you haven't generated this file yet. Please fix this.

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 11]
```diff
@@ -0,0 +1,952 @@
...
+In the Dockerfile, add the following:
+
+```dockerfile
+#Base image
+FROM node:lts-alpine
+
+#Install serve package
+npm i -g serve
+
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:873]: Missing the keyword `RUN`

Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:873]: Missing the keyword `RUN`

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md [hunk 12]
```diff
@@ -0,0 +1,952 @@
...
+services:        
+    flask-todos-api:
+        build: ./name_of_your_api_folder
+        ports: 
+            - 5000:5000
+
+    vue-todos-app:
+        build: ./name_of_your_client_folder
+        ports:
```
Comment (geoffrey45) [inline content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:916]: You used vue CLI to generate the Vue app. Replace the name you used earlier with `name_of_your_client_folder`

File: content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md (hunk 7 above)
Comment (mosesreigns) [reply content/articles/how-to-build-a-vue-app-with-flask-sqlite-backend-using-docker/index.md:205]: The app is using marshmallow to auto-generate the schemas

**Review by geoffrey45 — APPROVED** [review]

**Comment by hectorkambow** [comment]
Jumping on this one 👍

**Pull request merged**